File: src/utils/launchpad_utils.rs
use crate::types::launchpad::ProjectLink;
use std::collections::HashSet;

pub fn make_title(url: &str) -> String {
    match url::Url::parse(url) {
        Ok(parsed) => {
            let host = parsed.host_str().unwrap_or_default();
            if host.contains("crates.io") {
                return parsed
                    .path_segments()
                    .and_then(|mut segments| segments.next_back())
                    .unwrap_or(url)
                    .to_string();
            }

            if host.contains("github.com") {
                let segments = parsed
                    .path_segments()
                    .map(|seg| seg.collect::<Vec<_>>())
                    .unwrap_or_default();
                if segments.len() >= 2 {
                    return segments[1].to_string();
                }
            }

            host.to_string()
        }
        Err(_) => url.to_string(),
    }
}

pub fn host_type(url: &str) -> (&'static str, &'static str) {
    match url::Url::parse(url)
        .ok()
        .and_then(|parsed| parsed.host_str().map(str::to_string))
    {
        Some(host) if host.contains("github.com") => ("github.com", "GitHub"),
        Some(host) if host.contains("crates.io") => ("crates.io", "Crate"),
        Some(_) => ("external", "Link"),
        None => ("unknown", "Link"),
    }
}

pub fn filter_links(category: &str, links: &[ProjectLink], query: &str) -> Vec<ProjectLink> {
    let deduped_links = dedupe_links(links);
    let q = query.trim().to_lowercase();
    if q.is_empty() {
        return deduped_links;
    }

    deduped_links
        .iter()
        .filter(|link| {
            let title = make_title(&link.url).to_lowercase();
            title.contains(&q)
                || link.url.to_lowercase().contains(&q)
                || category.to_lowercase().contains(&q)
        })
        .cloned()
        .collect()
}

pub fn dedupe_links(links: &[ProjectLink]) -> Vec<ProjectLink> {
    let mut seen = HashSet::new();
    links
        .iter()
        .filter(|link| seen.insert(link.url.clone()))
        .cloned()
        .collect()
}

File: src/lib/launchpad-utils.ts
import type { ProjectLink } from '@/types/launchpad'

export function makeTitle(url: string): string {
  try {
    const parsed = new URL(url)
    const host = parsed.host
    const segments = parsed.pathname.split('/').filter(Boolean)

    if (host.includes('crates.io')) {
      return segments.at(-1) ?? url
    }

    if (host.includes('github.com') && segments.length >= 2) {
      return segments[1]
    }

    return host
  } catch {
    return url
  }
}

export function hostType(url: string): [host: string, kind: string] {
  try {
    const host = new URL(url).host
    if (host.includes('github.com')) return ['github.com', 'GitHub']
    if (host.includes('crates.io')) return ['crates.io', 'Crate']
    return ['external', 'Link']
  } catch {
    return ['unknown', 'Link']
  }
}

export function dedupeLinks(links: ProjectLink[]): ProjectLink[] {
  const seen = new Set<string>()
  return links.filter((link) => {
    if (seen.has(link.url)) return false
    seen.add(link.url)
    return true
  })
}

export function filterLinks(category: string, links: ProjectLink[], query: string): ProjectLink[] {
  const deduped = dedupeLinks(links)
  const q = query.trim().toLowerCase()
  if (!q) return deduped

  return deduped.filter((link) => {
    const title = makeTitle(link.url).toLowerCase()
    return (
      title.includes(q) ||
      link.url.toLowerCase().includes(q) ||
      category.toLowerCase().includes(q)
    )
  })
}
